### PR TITLE
Dual embedding save

### PR DESCRIPTION
**Pull request by tzufgoogle**
1. Saves a dictionary that maps cellids to embedding
2. Get text embedding

### CODE REVIEW

File: cabby/data/extract.py [hunk 1]
```diff
@@ -31,7 +31,7 @@
 
 def get_wikigeo_data(
     wikidata_items: Sequence[wdi.WikidataEntity]
-) -> Sequence[wikigeo.WikigeoEntity]:
```
Comment (sayalikulkarni) [inline cabby/data/extract.py:34]: may be separate this out from this PR to avoid conflicts with the other in-prgress ones?

File: cabby/model/text/dual_encoder/train.py [hunk 2]
```diff
@@ -30,164 +30,195 @@
...
+    total = 0
+    loss_val_total = 0
+
+    for batch in self.valid_loader:
+      text = {key: val.to(self.device) for key, val in batch['text'].items()}
+      cellids = batch['cellid'].float().to(self.device)
+      neighbor_cells = batch['neighbor_cells'].float().to(self.device) 
+      far_cells = batch['far_cells'].float().to(self.device)
+
+      cellids = cellids.float().to(self.device)
```
Comment (sayalikulkarni) [inline cabby/model/text/dual_encoder/train.py:80]: this should be int?

File: cabby/model/text/dual_encoder/train.py [hunk 3]
```diff
@@ -30,164 +30,195 @@
...
+    text_embedding_neighbor, cellid_embedding = self.model(text, neighbor_cells)
+    loss_neighbor = criterion(text_embedding_neighbor, 
+    cellid_embedding, target_neighbor)
+
+    # Far cellid.
+    target_far = -1*torch.ones(cellids.shape[0]).to(self.device)
+    text_embedding_far, cellid_embedding = self.model(text, far_cells)
+    loss_far = criterion(text_embedding_far, cellid_embedding, target_far)
+
+    loss = loss_cellid + loss_neighbor + loss_far
```
Comment (sayalikulkarni) [inline cabby/model/text/dual_encoder/train.py:137]: maybe add a positive loss weight? 
_num_negatives = 2
_positive_weight = 0.7
loss = (_positive_weight * loss_cellid) + (1-_positive_weight) * (loss_neighbor + loss_far) / _num_negatives

File: cabby/model/text/dual_encoder/train.py (hunk 2 above)
Comment (tzufgoogle) [reply cabby/model/text/dual_encoder/train.py:80]: No becuse it is a tensor type

File: cabby/model/text/dual_encoder/train.py (hunk 3 above)
Comment (tzufgoogle) [reply cabby/model/text/dual_encoder/train.py:137]: I tried that it did not improve

File: cabby/data/extract.py (hunk 1 above)
Comment (sayalikulkarni) [reply cabby/data/extract.py]: should be covered in merge

File: cabby/model/text/dual_encoder/train.py (hunk 2 above)
Comment (sayalikulkarni) [reply cabby/model/text/dual_encoder/train.py:80]: should be covered in merge

File: cabby/model/text/dual_encoder/train.py (hunk 3 above)
Comment (sayalikulkarni) [reply cabby/model/text/dual_encoder/train.py:137]: ack! good to know! do you have some place to track these variations you have been trying? Would be good to know what worked, what didnt etc. We usually keep a sheet to track the experiments for specific models e.g. see go/mage-geocoding-data 'experiments - 2020' sheet.

**Review by sayalikulkarni — APPROVED** [review]
minor suggestion. looks good to go. will sync offline on follow ups.
